library/std/src/os/windows/io/handle.rs

```diff
@@ -0,0 +1,427 @@
+///
+/// [`CloseHandle`]: https://docs.microsoft.com/en-us/windows/win32/api/handleapi/nf-handleapi-closehandle
+/// [`RegCloseKey`]: https://docs.microsoft.com/en-us/windows/win32/api/winreg/nf-winreg-regclosekey
+///
+/// [here]: https://devblogs.microsoft.com/oldnewthing/20040302-00/?p=40443
+#[repr(transparent)]
+#[unstable(feature = "io_safety", issue = "87074")]
+pub struct OwnedHandle {
+    handle: NonNull<c_void>,
+}
+
+/// Similar to `Option<OwnedHandle>`, but intended for use in FFI interfaces
+/// where `INVALID_HANDLE_VALUE` is used as the sentry value, and null values
+/// are not used at all, such as in the return value of `CreateFileW`.
+///
+/// The main thing you can do with an `OptionFileHandle` is to convert it into
+/// an `OwnedHandle` using its [`TryFrom`] implementation, and this conversion
+/// takes care of the check for `INVALID_HANDLE_VALUE`.
+///
+/// If this holds an owned handle, it closes the handle on drop.
+///
+/// This uses `repr(transparent)` and has the representation of a host handle,
+/// so it can be used in FFI in places where a non-null handle is passed as a
+/// consumed argument or returned as an owned value, or it is
+/// `INVALID_HANDLE_VALUE` indicating an error or an otherwise absent value.
+#[repr(transparent)]
+#[unstable(feature = "io_safety", issue = "87074")]
+pub struct OptionFileHandle {
+    handle: RawHandle,
+}
+
+// The Windows [`HANDLE`] type may be transferred across and shared between
+// thread boundaries (despite containing a `*mut void`, which in general isn't
+// `Send` or `Sync`).
+//
+// [`HANDLE`]: std::os::windows::raw::HANDLE
+unsafe impl Send for OwnedHandle {}
+unsafe impl Send for OptionFileHandle {}
+unsafe impl Send for BorrowedHandle<'_> {}
+unsafe impl Sync for OwnedHandle {}
+unsafe impl Sync for OptionFileHandle {}
+unsafe impl Sync for BorrowedHandle<'_> {}
+
+impl BorrowedHandle<'_> {
+    /// Return a `BorrowedHandle` holding the given raw handle.
+    ///
+    /// # Safety
+    ///
+    /// The resource pointed to by `handle` must be a valid open handle, it
+    /// must remain open for the duration of the returned `BorrowedHandle`, and
+    /// it must not be null.
+    ///
+    /// Note that it *may* have the value `INVALID_HANDLE_VALUE` (-1), which is
+    /// sometimes a valid handle value. See [here] for the full story.
+    ///
+    /// [here]: https://devblogs.microsoft.com/oldnewthing/20040302-00/?p=40443
+    #[inline]
+    #[unstable(feature = "io_safety", issue = "87074")]
+    pub unsafe fn borrow_raw_handle(handle: RawHandle) -> Self {
+        assert!(!handle.is_null());
+        Self { handle: NonNull::new_unchecked(handle), _phantom: PhantomData }
+    }
+}
+
+impl OptionFileHandle {
+    /// Return an empty `OptionFileHandle` with no resource.
+    #[inline]
+    #[unstable(feature = "io_safety", issue = "87074")]
+    pub const fn none() -> Self {
+        Self { handle: c::INVALID_HANDLE_VALUE }
+    }
+}
```

Comment:
I think this should be dropped, in favor of people calling the TryFrom implementation.
```suggestion
```